libs/failsafe-controller/src/main/java/net/jodah/failsafe/internal/actions/ActionRegistry.java

```diff
@@ -0,0 +1,340 @@
+import org.apache.commons.lang.Validate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The action registry class is used to track failsafe actions registered with a controller.
+ *
+ * @param <R> the result type
+ */
+public class ActionRegistry<R> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ActionRegistry.class);
+
+  private static final String CR_AND_INDENT = "\r\n\t";
+
+  private final FailsafeController<R> controller;
+
+  private final Deque<Expectation> expectations = new LinkedList<>();
+
+  private final Deque<Expectation> processing = new LinkedList();
+
+  private AssertionError shutdownFailure = null;
+
+  public ActionRegistry(FailsafeController<R> controller) {
+    this.controller = controller;
+  }
+
+  /**
+   * Adds the specified action list as the next expected execution.
+   *
+   * @param actionList the list of expected actions for the next expected execution
+   */
+  public void add(ActionList<R> actionList) {
+    synchronized (controller) {
+      final Expectation expectation = new Expectation(expectations.size() + 1);
+
+      actionList.populate(expectation);
+      expectations.add(expectation);
+    }
+  }
+
+  /**
+   * Adds the specified action list as the next <code>count</code> expected executions.
+   *
+   * @param actionList the list of expected actions for the next expected executions
+   * @param count the number of expected executions to register the list of actions with
+   * @throws IllegalArgumentException if <code>count</code> is negative
+   */
+  public void add(ActionList<R> actionList, int count) {
+    Validate.isTrue(count >= 0, "count must be greater or equal than 0");
+    synchronized (controller) {
+      for (int i = 0; i < count; i++) {
+        final Expectation expectation = new Expectation(expectations.size() + 1);
+
+        actionList.populate(expectation);
+        expectations.add(expectation);
+      }
+    }
+  }
+
+  /**
+   * Shuts down this registry.
+   *
+   * @param failure the associated failure to throw back if the registry is used afterward
+   */
+  public void shutdown(AssertionError failure) {
+    synchronized (controller) {
+      this.shutdownFailure = failure;
+    }
+  }
+
+  /**
+   * Verifies if all expected actions that cannot be left incomplete have been completed.
+   *
+   * @throws AssertionError if there are recorded actions left to execute
+   */
+  public void verify() {
+    synchronized (controller) {
+      processing.forEach(Expectation::verify);
+      if (!expectations.isEmpty()) {
+        throw controller.setFailure(
+            new AssertionError(
+                "expecting "
+                    + (expectations.size() + processing.size())
+                    + " executions for '"
+                    + controller
+                    + "' but only "
+                    + processing.size()
+                    + " occurred; the following were not executed: \r\n\t"
```

Comment:
✏️ Did you mean to use your CR_AND_INDENT variable here? also this entire string might be cleaner if done as a string.format